File: LMS-database-schema-creation-1129469251547016005/src/lib/types/student.ts
export interface Student {
  id: string;
  studentId: string;
  name: string;
  email: string;
  avatar?: string;
  semester: number;
  grade: 10 | 11 | 12; // Tingkat kelas
  school: {
    name: string;
    province: string; // e.g., "Jawa Barat"
  };
  phone: string;
  class: string;
  major: string;
  enrollmentDate: string;
  status: "active" | "inactive" | "suspended";
  gpa: number;

  // Academic Info
  totalCredits: number;
  completedCredits: number;

  // Contact Info
  address?: string;
  parentName?: string;
  parentPhone?: string;

  // Stats
  attendanceRate: number;
  assignmentCompletion: number;
  quizAverage: number;

  // Risk Analysis
  riskStatus?: "Aman" | "Berisiko Sedang" | "Berisiko Tinggi";
  riskScore?: number;

  createdAt: string;
  updatedAt: string;
}

export interface StudentGrade {
  id: string;
  studentId: string;
  subject: string;
  type: string;
  title: string;
  score: number;
  maxScore: number; // e.g., A, B, C, etc.
  percentage: number;
  grade: string;
  submittedAt: string; // Optional remarks or comments
  gradedAt: string;
}

export interface StudentAttendance {
  id: string;
  studentId: string;
  date: string;
  status: "present" | "absent" | "late" | "excused";
  subject: string;
  checkInTime: string; // e.g., "09:00 AM"
  location: string; // e.g., "Room 101"
  notes: string; // Optional remarks or comments
}

export interface StudentActivity {
  id: string;
  studentId: string;
  type:
    | "login"
    | "logout"
    | "quiz_complete"
    | "forum_post"
    | "video_join"
    | "assignment_submit";
  timestamp: string;
  title: string; // e.g., "Quiz 1 Completed"
  description: string; // e.g., "Logged in from IP
  metadata?: Record<string, any>; // Additional data related to the activity
}

File: backend_rs/src/models/student.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use chrono::{DateTime, Utc};
use std::fmt;

#[derive(Debug, Serialize, Deserialize, Clone, sqlx::Type)]
#[sqlx(type_name = "status", rename_all = "snake_case")]
pub enum StudentStatus {
    Active,
    Inactive,
    Suspended,
}

impl fmt::Display for StudentStatus {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            StudentStatus::Active => write!(f, "active"),
            StudentStatus::Inactive => write!(f, "inactive"),
            StudentStatus::Suspended => write!(f, "suspended"),
        }
    }
}

#[derive(Debug, Serialize, Deserialize, FromRow, Clone)]
pub struct Student {
    pub id: String,
    pub student_id: String, // NIM
    pub name: String,
    pub email: String,
    pub avatar: Option<String>,
    pub semester: Option<i32>,
    pub grade: Option<i32>,
    pub school_name: Option<String>,
    pub school_province: Option<String>,
    pub phone: Option<String>,
    pub class_name: Option<String>,
    pub major: Option<String>,
    pub enrollment_date: Option<DateTime<Utc>>,
    pub status: Option<String>, // simplified from enum for DB compatibility if needed, or use sqlx::Type
    pub gpa: Option<f64>, // Decimal(4,2) mapped to f64
    pub total_credits: Option<i32>,
    pub completed_credits: Option<i32>,
    pub address: Option<String>,
    pub parent_name: Option<String>,
    pub parent_phone: Option<String>,
    pub attendance_rate: Option<f64>,
    pub assignment_completion: Option<f64>,
    pub quiz_average: Option<f64>,
    pub risk_status: Option<String>,
    pub risk_score: Option<f64>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct StudentRiskUpdate {
    pub risk_status: String,
    pub risk_score: f64,
}
